src/librbd/mirror_snapshot/CreateRequest.h

```diff
@@ -0,0 +1,113 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_LIBRBD_MIRROR_SNAPSHOT_CREATE_REQUEST_H
+#define CEPH_LIBRBD_MIRROR_SNAPSHOT_CREATE_REQUEST_H
+
+#include "include/buffer.h"
+#include "cls/rbd/cls_rbd_types.h"
+
+#include <string>
+#include <set>
+
+struct Context;
+
+namespace librbd {
+
+struct ImageCtx;
+
+namespace mirror_snapshot {
+
+template <typename ImageCtxT = librbd::ImageCtx>
+class CreateRequest {
+public:
+  static CreateRequest *create(ImageCtxT *image_ctx,
+                               cls::rbd::MirrorSnapshotType type,
+                               const std::string &mirror_uuid,
+                               uint64_t remote_snap_id,
+                               uint64_t *snap_id, Context *on_finish) {
+    return new CreateRequest(image_ctx, type, mirror_uuid, remote_snap_id,
+                             snap_id, on_finish);
+  }
+
+  CreateRequest(ImageCtxT *image_ctx, cls::rbd::MirrorSnapshotType type,
+                const std::string &mirror_uuid, uint64_t remote_snap_id,
+                uint64_t *snap_id, Context *on_finish)
+    : m_image_ctx(image_ctx), m_type(type), m_mirror_uuid(mirror_uuid),
+      m_remote_snap_id(remote_snap_id), m_snap_id(snap_id),
+      m_on_finish(on_finish) {
+  }
+
+  void send();
+
+private:
+  /**
+   * @verbatim
+   *
+   * <start>
+   *    |
+   *    v
+   * REFRESH_IMAGE
+   *    |
+   *    v
+   * GET_MIRROR_IMAGE
+   *    |
+   *    v
+   * GET_MIRROR_UUID (skip if not needed)
+   *    |
+   *    v
+   * GET_MIRROR_PEERS (skip if not needed)
+   *    |
+   *    v
+   * UNLINK_PEER (skip if not needed,
```

Comment:
Is there a race condition here? Should the snapshot be created first prior to pruning?